File: app/providers.py
"""LLM Provider Registry and Factory for multi-provider support."""

from dataclasses import dataclass
from typing import Optional
from langchain_core.language_models.chat_models import BaseChatModel


@dataclass
class ProviderConfig:
    """Configuration for an LLM provider."""
    package: str
    class_name: str
    env_key: str
    default_model: str
    emoji: str
    description: str


PROVIDER_REGISTRY: dict[str, ProviderConfig] = {
    "google_gemini": ProviderConfig(
        package="langchain_google_genai",
        class_name="ChatGoogleGenerativeAI",
        env_key="GOOGLE_API_KEY",
        default_model="gemini-2.0-flash",
        emoji="💎",
        description="Google Gemini models"
    ),
    "cohere": ProviderConfig(
        package="langchain_cohere",
        class_name="ChatCohere",
        env_key="COHERE_API_KEY",
        default_model="command-r-plus",
        emoji="🔮",
        description="Cohere Command models"
    ),
    "mistral": ProviderConfig(
        package="langchain_mistralai",
        class_name="ChatMistralAI",
        env_key="MISTRAL_API_KEY",
        default_model="mistral-large-latest",
        emoji="🌬️",
        description="Mistral AI models"
    ),
    "groq": ProviderConfig(
        package="langchain_groq",
        class_name="ChatGroq",
        env_key="GROQ_API_KEY",
        default_model="llama-3.3-70b-versatile",
        emoji="⚡",
        description="Groq fast inference"
    ),
    "nvidia": ProviderConfig(
        package="langchain_nvidia_ai_endpoints",
        class_name="ChatNVIDIA",
        env_key="NVIDIA_API_KEY",
        default_model="meta/llama-3.1-8b-instruct",
        emoji="🟢",
        description="NVIDIA NIM endpoints"
    ),
    "cerebras": ProviderConfig(
        package="langchain_cerebras",
        class_name="ChatCerebras",
        env_key="CEREBRAS_API_KEY",
        default_model="qwen-3-235b-a22b-instruct-2507",
        emoji="🧠",
        description="Cerebras ultra-fast inference"
    ),
    "huggingface": ProviderConfig(
        package="langchain_huggingface",
        class_name="ChatHuggingFace",
        env_key="HUGGINGFACE_API_KEY",
        default_model="Qwen/Qwen2.5-Coder-32B-Instruct",
        emoji="🤗",
        description="Hugging Face Inference API"
    ),
    "openrouter": ProviderConfig(
        package="langchain_openrouter",
        class_name="ChatOpenRouter",
        env_key="OPENROUTER_API_KEY",
        default_model="meta-llama/llama-3-70b-instruct",
        emoji="🌐",
        description="OpenRouter multi-model gateway"
    ),
    "github_models": ProviderConfig(
        package="langchain_openai",
        class_name="ChatOpenAI",
        env_key="GITHUB_MODELS_TOKEN",
        default_model="gpt-4o",
        emoji="🐙",
        description="GitHub Models (via OpenAI-compatible API)"
    ),
}


def get_llm(provider: str, model: Optional[str] = None, api_key: Optional[str] = None) -> BaseChatModel:
    """
    Factory function to instantiate an LLM from the provider registry.
    
    Args:
        provider: Provider name (e.g., 'groq', 'cerebras', 'nvidia')
        model: Model name (optional, uses provider default if not specified)
        api_key: API key (optional, reads from env if not specified)
    
    Returns:
        A BaseChatModel instance configured for the specified provider
    
    Raises:
        ValueError: If provider is not in the registry
        ImportError: If the required package is not installed
    """
    if provider not in PROVIDER_REGISTRY:
        available = ", ".join(PROVIDER_REGISTRY.keys())
        raise ValueError(f"Unknown provider '{provider}'. Available: {available}")
    
    config = PROVIDER_REGISTRY[provider]
    
    # Get API key from env if not provided
    if api_key is None:
        import os
        api_key = os.environ.get(config.env_key)
        if api_key is None:
            raise ValueError(f"API key not found. Set {config.env_key} environment variable.")
    
    # Import the provider class dynamically
    try:
        module = __import__(config.package, fromlist=[config.class_name])
        chat_class = getattr(module, config.class_name)
    except ImportError as e:
        raise ImportError(
            f"Package '{config.package}' not installed. "
            f"Run: pip install {config.package}"
        ) from e
    
    # Build kwargs for the chat model
    kwargs = {"api_key": api_key}
    
    if model:
        kwargs["model"] = model
    else:
        kwargs["model"] = config.default_model
    
    # Special handling for GitHub Models (uses custom base_url)
    if provider == "github_models":
        kwargs["base_url"] = "https://models.inference.ai.azure.com"
    
    return chat_class(**kwargs)


def list_providers() -> dict[str, ProviderConfig]:
    """
    Returns the provider registry for display purposes.
    
    Returns:
        Dict mapping provider names to their configurations
    """
    return PROVIDER_REGISTRY.copy()

File: src/providers.js
/**
 * Multi-provider LLM factory.
 * All providers expose OpenAI-compatible /v1/chat/completions endpoints.
 */

const PROVIDERS = {
	cerebras: {
		baseUrl: "https://api.cerebras.ai/v1",
		envKey: "CEREBRAS_API_KEY",
		defaultModel: "qwen-3-235b-a22b-instruct-2507",
	},
	groq: {
		baseUrl: "https://api.groq.com/openai/v1",
		envKey: "GROQ_API_KEY",
		defaultModel: "llama-3.3-70b-versatile",
	},
	mistral: {
		baseUrl: "https://api.mistral.ai/v1",
		envKey: "MISTRAL_API_KEY",
		defaultModel: "mistral-large-latest",
	},
	google_gemini: {
		baseUrl:
			"https://generativelanguage.googleapis.com/v1beta/openai",
		envKey: "GOOGLE_API_KEY",
		defaultModel: "gemini-2.0-flash",
	},
	nvidia: {
		baseUrl: "https://integrate.api.nvidia.com/v1",
		envKey: "NVIDIA_API_KEY",
		defaultModel: "meta/llama-3.1-8b-instruct",
	},
	cohere: {
		baseUrl: "https://api.cohere.com/v2",
		envKey: "COHERE_API_KEY",
		defaultModel: "command-r-plus",
	},
	huggingface: {
		baseUrl: "https://api-inference.huggingface.co/v1",
		envKey: "HUGGINGFACE_API_KEY",
		defaultModel: "Qwen/Qwen2.5-Coder-32B-Instruct",
	},
	openrouter: {
		baseUrl: "https://openrouter.ai/api/v1",
		envKey: "OPENROUTER_API_KEY",
		defaultModel: "meta-llama/llama-3-70b-instruct",
	},
	github_models: {
		baseUrl: "https://models.inference.ai.azure.com",
		envKey: "GITHUB_MODELS_TOKEN",
		defaultModel: "gpt-4o",
	},
};

/**
 * Send a chat completion request to the configured provider.
 * @param {Env} env - Worker env bindings
 * @param {Array<{role: string, content: string}>} messages - Chat messages
 * @returns {Promise<string>} The assistant's response content
 */
export async function chatCompletion(env, messages) {
	const providerName = env.AI_PROVIDER || "cerebras";
	const config = PROVIDERS[providerName];

	if (!config) {
		const available = Object.keys(PROVIDERS).join(", ");
		throw new Error(
			`Unknown provider '${providerName}'. Available: ${available}`,
		);
	}

	const apiKey = env[config.envKey];
	if (!apiKey) {
		throw new Error(
			`API key not found. Set ${config.envKey} environment variable.`,
		);
	}

	const model = env.AI_MODEL || config.defaultModel;
	const url = `${config.baseUrl}/chat/completions`;

	const response = await fetch(url, {
		method: "POST",
		headers: {
			"Content-Type": "application/json",
			Authorization: `Bearer ${apiKey}`,
		},
		body: JSON.stringify({
			model,
			messages,
			max_tokens: 4096,
		}),
	});

	if (!response.ok) {
		const text = await response.text();
		throw new Error(
			`LLM API error (${providerName}): ${response.status} ${text}`,
		);
	}

	const data = await response.json();
	return data.choices[0].message.content;
}

/**
 * Return the provider registry.
 * @returns {Record<string, object>}
 */
export function listProviders() {
	return { ...PROVIDERS };
}
